compiler/z/codegen/ControlFlowEvaluator.cpp

```diff
@@ -323,52 +324,174 @@ generateS390Compare(TR::Node * node, TR::CodeGenerator * cg, TR::InstOpCode::Mne
    return branchOpCond;
    }
 
-static TR::Register*
-xmaxxminHelper(TR::Node* node, TR::CodeGenerator* cg, TR::InstOpCode::Mnemonic compareRROp, TR::InstOpCode::S390BranchCondition branchCond, TR::InstOpCode::Mnemonic moveRROp)
+TR::Register *
+OMR::Z::TreeEvaluator::xmaxxminHelper(TR::Node * node, TR::CodeGenerator * cg)
    {
-   TR::Node* lhsNode = node->getChild(0);
-   TR::Node* rhsNode = node->getChild(1);
-
-   TR::Register* lhsReg = cg->gprClobberEvaluate(lhsNode);
-   TR::Register* rhsReg = cg->evaluate(rhsNode);
+   TR::Register * operand1 = cg->gprClobberEvaluate(node->getChild(0));
+   TR::Register * operand2 = cg->evaluate(node->getChild(1));
 
    TR::LabelSymbol* cFlowRegionStart = generateLabelSymbol(cg);
-   TR::LabelSymbol* cFlowRegionEnd = generateLabelSymbol(cg);
-
    generateS390LabelInstruction(cg, TR::InstOpCode::label, node, cFlowRegionStart);
    cFlowRegionStart->setStartInternalControlFlow();
 
-   generateRREInstruction(cg, compareRROp, node, lhsReg, rhsReg);
-   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, branchCond, node, cFlowRegionEnd);
-
-   //Check for NaN operands for float and double
-   if (node->getOpCode().isFloatingPoint())
+   TR::InstOpCode::Mnemonic compareRROp = TR::InstOpCode::NOP;
+   TR::InstOpCode::S390BranchCondition returnFirstArgCond = TR::InstOpCode::COND_NOP;
+   bool isMaxOp = node->getOpCode().isMax();
+   bool isFloatingPointOp = node->getOpCode().isFloatingPoint();
+   if (isFloatingPointOp)
       {
+      compareRROp = node->getOpCode().isDouble() ? TR::InstOpCode::CDBR : TR::InstOpCode::CEBR;
+      returnFirstArgCond = isMaxOp ? TR::InstOpCode::COND_MASK2 : TR::InstOpCode::COND_MASK4;
+      }
+   else
+      {
+      TR_ASSERT_FATAL(node->getOpCode().isInt() || node->getOpCode().isLong(), "invalid operand type");
+      compareRROp = node->getOpCode().isLong() ? TR::InstOpCode::CGR : TR::InstOpCode::CR;
+      returnFirstArgCond = isMaxOp ? TR::InstOpCode::COND_BHR : TR::InstOpCode::COND_BLR;
+      }
+
+   TR::LabelSymbol* cFlowRegionEnd = generateLabelSymbol(cg);
+   TR::LabelSymbol* swapValues = generateLabelSymbol(cg);
+   generateRREInstruction(cg, compareRROp, node, operand1, operand2);
+   generateS390BranchInstruction(cg, TR::InstOpCode::BRC, returnFirstArgCond, node, cFlowRegionEnd);
+   if (isFloatingPointOp)
+      {
+      /* Check for NaN operands for float and double
+       * Support float and double +0/-0 comparisons adhering to IEEE 754 standard
+       * Checking if operands are equal, then branching to equalRegion, otherwise
+       * fall through for NaN case handling
+       */
```

Comment:
```suggestion
      /**
       * Check for NaN operands for float and double
       * Support float and double +0/-0 comparisons adhering to IEEE 754 standard
       * Checking if operands are equal, then branching to equalRegion, otherwise
       * fall through for NaN case handling
       */
```